npe2/_plugin_manager.py

```diff
@@ -52,105 +54,220 @@ def addi(self, begin: int, end: int, data: T) -> None:
+            # 1. writers with no file extensions (like directory writers) go last
+            no_ext = len(writer.filename_extensions) == 0
+
+            # 2. more "specific" writers first
+            nbounds = sum(not c.is_zero() for c in writer.layer_type_constraints())
+
+            # 3. then sort by the number of listed extensions
+            #    (empty set of extensions goes last)
+            ext_len = len(writer.filename_extensions)
+
+            # 4. finally group related extensions together
+            exts = writer.filename_extensions
+            return (no_ext, nbounds, ext_len, exts)
+
+        yield from sorted(
+            candidates,
+            key=_writer_key,
+        )
+
+
+class PluginManagerEvents(SignalGroup):
+    plugins_registered = Signal(
+        set,
+        description="Emitted with a set of PluginManifest instances "
+        "whenever new plugins are registered. 'Registered' means that a "
+        "manifest has been provided or discovered.",
+    )
+    activation_changed = Signal(
+        set,
+        set,
+        description="Emitted with two arguments: a set of plugin "
+        "names that were activated, and a set of names that were "
+        "deactivated. 'Activated' means the plugin has been *imported*, it's "
+        "`on_activate` function was called.",
+    )
+    enablement_changed = Signal(
+        set,
+        set,
+        description="Emitted with two arguments: a set of plugin names "
+        "that were enabled, and a set of names that were "
+        "disabled. 'Disabled' means the plugin remains installed, but it "
+        "cannot be activated, and its contributions will not be indexed.",
+    )
 
 
 class PluginManager:
     __instance: Optional[PluginManager] = None  # a global instance
     _contrib: _ContributionsIndex
+    events: PluginManagerEvents
 
-    def __init__(self, reg: Optional[CommandRegistry] = None) -> None:
+    def __init__(
+        self, *, disable: Iterable[str] = (), reg: Optional[CommandRegistry] = None
+    ) -> None:
+        self._disabled_plugins: Set[PluginName] = set(disable)
         self._command_registry = reg or CommandRegistry()
         self._contexts: Dict[PluginName, PluginContext] = {}
+        self._contrib = _ContributionsIndex()
         self._manifests: Dict[PluginName, PluginManifest] = {}
-        self.discover()  # TODO: should we be immediately discovering?
+        self.events = PluginManagerEvents(self)
+
+    @classmethod
+    def instance(cls) -> PluginManager:
+        """Singleton instance."""
+        if cls.__instance is None:
+            cls.__instance = cls()
+        return cls.__instance
 
     @property
     def commands(self) -> CommandRegistry:
         return self._command_registry
 
-    def discover(self, paths: Sequence[str] = ()) -> None:
+    # Discovery, activation, enablement
+
+    def discover(self, paths: Sequence[str] = (), clear=False) -> None:
```

Comment:
I believe technically the super class also allows Path in addition to strings.

Reply:
good point.  I think we want to remove discover from PluginManifest altogether and move that logic here.  so might do in a followup pr